File: robot-factory/python/tests/fake_part_supplier.py
from __future__ import annotations

from robot_factory import Money, PartOption, PartQuote, PartType, PurchasedPart


class FakePartSupplier:
    def __init__(self, name: str) -> None:
        self._name = name
        self._catalog: dict[tuple[PartType, PartOption], Money] = {}
        self._purchase_log: list[PurchasedPart] = []

    @property
    def name(self) -> str:
        return self._name

    @property
    def purchase_log(self) -> list[PurchasedPart]:
        return list(self._purchase_log)

    def with_part(
        self, part_type: PartType, option: PartOption, price: Money
    ) -> FakePartSupplier:
        self._catalog[(part_type, option)] = price
        return self

    def get_quote(
        self, part_type: PartType, option: PartOption
    ) -> PartQuote | None:
        price = self._catalog.get((part_type, option))
        if price is None:
            return None
        return PartQuote(
            type=part_type, option=option, price=price, supplier_name=self._name
        )

    def purchase(
        self, part_type: PartType, option: PartOption
    ) -> PurchasedPart:
        price = self._catalog[(part_type, option)]
        part = PurchasedPart(
            type=part_type, option=option, price=price, supplier_name=self._name
        )
        self._purchase_log.append(part)
        return part

File: robot-factory/typescript/tests/FakePartSupplier.ts
import { Money } from '../src/Money.js';
import type { PartOption } from '../src/PartOption.js';
import type { PartQuote } from '../src/PartQuote.js';
import type { PartSupplier } from '../src/PartSupplier.js';
import type { PartType } from '../src/PartType.js';
import type { PurchasedPart } from '../src/PurchasedPart.js';

export class FakePartSupplier implements PartSupplier {
  private readonly catalog = new Map<string, Money>();
  private readonly _purchaseLog: PurchasedPart[] = [];

  constructor(public readonly name: string) {}

  get purchaseLog(): readonly PurchasedPart[] { return this._purchaseLog; }

  withPart(type: PartType, option: PartOption, price: Money): this {
    this.catalog.set(`${type}:${option}`, price);
    return this;
  }

  getQuote(type: PartType, option: PartOption): PartQuote | null {
    const price = this.catalog.get(`${type}:${option}`);
    return price ? { type, option, price, supplierName: this.name } : null;
  }

  purchase(type: PartType, option: PartOption): PurchasedPart {
    const price = this.catalog.get(`${type}:${option}`)!;
    const part: PurchasedPart = { type, option, price, supplierName: this.name };
    this._purchaseLog.push(part);
    return part;
  }
}
